### PR TITLE
Use std::string instead of error-prone C-strings

### PR DESCRIPTION
**Pull request by 3v1n0**
This requires https://github.com/3v1n0/snapcraft-preload/commits/c%2B%2B-rewrite branch... We remove the usage of `char*` strings when we handle them, by converting in `std::string`'s and using `std::vector` when appropriate.

### CODE REVIEW

**Comment by 3v1n0** [comment]
@kyrofa hey, please once you've re-reviewed the c++-rewrite branch, give a look to this too. :-)

**Comment by kyrofa** [comment]
Should this be rebased now that the prereq has been merged?

**Comment by kyrofa** [comment]
Oh it hasn't, haha! Ignore me...

**Comment by sergiusens** [comment]
it has now!

**Comment by 3v1n0** [comment]
@kyrofa ok, rebased... It should be good for review now

File: preload.cpp [hunk 1]
```diff
@@ -36,226 +38,220 @@
...
 using compar_function_t = int (*)(const dirent_t **, const dirent_t **);
-} // unnamed namespace
 
-static void constructor() __attribute__((constructor));
+using socket_action_t = int (*) (int, const struct sockaddr *, socklen_t);
+using execve_t = int (*) (const char *, char *const[], char *const[]);
 
-static char *
-getenvdup (const char *varname, size_t *envlen)
+inline std::string
```
Comment (kyrofa) [inline preload.cpp]: I'm curious about your justification for inlining, here. Seems like potentially premature optimization to me. Why not just leave it up to the compiler? Same comment for all of these inline functions, here.

File: preload.cpp [hunk 2]
```diff
@@ -36,226 +38,220 @@
...
-        }
-        return NULL;
-    }
-    else {
-        if (envlen) {
-            *envlen = strlen (envvar);
-            return strndup (envvar, *envlen);
-        }
+    char *envvar = secure_getenv (varname.c_str());
+    return envvar ? envvar : "";
```
Comment (kyrofa) [inline preload.cpp]: I suggest avoiding such constructs. A simple `if` is more clear.

**Pull request merged**